evm_arithmetization/src/cpu/kernel/interpreter.rs

```diff
@@ -575,33 +472,53 @@ impl<'a, F: Field> Interpreter<'a, F> {
         }
     }
 
-    pub(crate) fn get_trie_data(&self) -> &[U256] {
-        &self.generation_state.memory.contexts[0].segments[Segment::TrieData.unscale()].content
+    pub(crate) fn get_trie_data(&self) -> Vec<U256> {
+        self.generation_state.memory.contexts[0].segments[Segment::TrieData.unscale()]
+            .content
+            .iter()
+            .filter_map(|&elt| elt)
+            .collect::<Vec<_>>()
     }
 
-    pub(crate) fn get_trie_data_mut(&mut self) -> &mut Vec<U256> {
+    pub(crate) fn get_trie_data_mut(&mut self) -> &mut Vec<Option<U256>> {
         &mut self.generation_state.memory.contexts[0].segments[Segment::TrieData.unscale()].content
     }
 
     pub(crate) fn get_memory_segment(&self, segment: Segment) -> Vec<U256> {
-        self.generation_state.memory.contexts[0].segments[segment.unscale()]
-            .content
-            .clone()
+        if self.preinitialized_segments.contains_key(&segment) {
+            let total_len = self.generation_state.memory.contexts[0].segments[segment.unscale()]
+                .content
+                .len();
+            let get_vals = |opt_vals: &[Option<U256>]| {
+                opt_vals
+                    .iter()
+                    .map(|&elt| match elt {
+                        Some(val) => val,
+                        None => U256::zero(),
+                    })
+                    .collect::<Vec<U256>>()
```

Comment:
There's a lot of this stuff going on in the PR, a consequence of using `Option<U256>`. Are we sure we can't avoid wrapping values in `Option`? Can we just use `0` or `U256::MAX` or some other trickery?

Reply:
The problem is that we need to know whether a certain memory spot was set or not. But memory values can take any U256 value, so unfortunately, I think we do need the `Option<U256>`...